### PR TITLE
CUMULUS-812: Add integration tests for executions started from queues with a maximum

### PR DESCRIPTION
**Pull request by markdboyd**
**Summary:** Summary of changes

Addresses [CUMULUS-812: As a DAAC operator, I want to control the maximum executions for a queue so that urgent processing needs are not queued behind less urgent needs](https://bugs.earthdata.nasa.gov/browse/CUMULUS-812)

## Changes

* Add integration tests to ensure that the number of running executions started from a queue never exceeds the maximum allowed number for that queue
* Update unit tests for sf-scheduler, queue-granules, and queue-pdrs to ensure `meta.queueExecutionLimits` from message template is not overwritten

## PR Checklist

- [ ] Update CHANGELOG
- [x] Unit tests
- [x] Integration tests

### CODE REVIEW

File: example/spec/parallel/startSF/startSfSpec.js [hunk 1]
```diff
@@ -274,94 +238,25 @@ describe('the sf-starter lambda function', () => {
       ]);
     });
 
-    it('queue-granules returns the correct amount of queued executions', async () => {
```
Comment (markdboyd) [inline example/spec/parallel/startSF/startSfSpec.js]: I decided that queuing the executions via queue-granules wasn't really essential for what this test is trying to prove. This test should only care that for any execution messages queued with the necessary values, the executions are limited to the max for the queue.

This is also why I updated the unit tests for the various lambdas that add messages to the queues (queue-granules, queue-pdrs, sf-scheduler) to ensure that they all preserve the necessary message attributes for the execution throttling to happen.

File: example/spec/parallel/startSF/startSfSpec.js [hunk 2]
```diff
@@ -141,15 +168,14 @@ describe('the sf-starter lambda function', () => {
     });
 
     it('consumes the messages', async () => {
-      const response = await lambda().invoke({
+      const { Payload } = await lambda().invoke({
```
Comment (markdboyd) [inline example/spec/parallel/startSF/startSfSpec.js]: You'll notice that I opted to not change the test expectation on line 192 of this file:

https://github.com/nasa/cumulus/pull/1062/files#diff-89bd6949a4ceda77cbb0a67687670574R192

Ultimately I just didn't like any of the possible solutions. And given that this test is not new for this PR and I'm not aware of any reported failures from CI, it seemed like it might be best to leave it alone.

File: example/spec/parallel/startSF/startSfSpec.js [hunk 3]
```diff
@@ -199,72 +211,24 @@ describe('the sf-starter lambda function', () => {
...
-
       await Promise.all([
-        deleteS3Object(config.bucket, templateKey),
         sqs().deleteQueue({
           QueueUrl: maxQueueUrl
         }).promise(),
-        deleteCollections(config.stackName, config.bucket, [collection]),
-        deleteStateMachine,
+        sfn().deleteStateMachine({ stateMachineArn: waitPassSfArn }).promise(),
         dynamodbDocClient().delete({
```
Comment (markdboyd) [inline example/spec/parallel/startSF/startSfSpec.js]: This does delete the semaphore record, but because this test doesn't wait for the executions to finish, once they do finish they call `semaphore.down()`, which will always create the record if it doesn't exist before decrementing. Without making the test wait for all executions to finish, which seems wasteful, I'm not sure what to do here, but I don't love that the test doesn't fully clean up after itself.

File: example/spec/parallel/startSF/startSfSpec.js [hunk 4]
```diff
@@ -141,15 +168,14 @@ describe('the sf-starter lambda function', () => {
...
         FunctionName: sfStarterName,
         InvocationType: 'RequestResponse',
         Payload: JSON.stringify({
           queueUrl,
           messageLimit: testMessageLimit
         })
       }).promise();
-      const { Payload } = response;
       messagesConsumed = parseInt(Payload, 10);
       expect(messagesConsumed).toBeGreaterThan(0);
```
Comment (ifestus) [inline example/spec/parallel/startSF/startSfSpec.js]: I don't understand why this expression is checking for `> 0`. Should this be `>= maxExecutions`?

File: example/spec/parallel/startSF/startSfSpec.js [hunk 5]
```diff
@@ -76,45 +63,79 @@ const waitPassSfParams = {
...
 
-const sfStarterName = `${config.stackName}-sqs2sf`;
+async function sendStartSfMessages({
+  numOfMessages,
+  queueMaxExecutions,
+  queueName,
+  queueUrl,
+  workflowArn
+}) {
+  const message = {
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply example/spec/parallel/startSF/startSfSpec.js]: Oh my. Good catch

File: example/spec/parallel/startSF/startSfSpec.js (hunk 4 above)
Comment (markdboyd) [reply example/spec/parallel/startSF/startSfSpec.js]: This test is not for a queue with a maximum number of executions defined, but a regular queue. So the amount of messages it would consume (up to the message limit passed to the Lambda) is unpredictable.

File: example/spec/parallel/startSF/startSfSpec.js [hunk 6]
```diff
@@ -274,94 +238,25 @@ describe('the sf-starter lambda function', () => {
...
       }).promise();
-
-      console.log(response);
-      const { Payload } = response;
-
       messagesConsumed = parseInt(Payload, 10);
-      console.log('messages consumed', messagesConsumed);
       // Can't test that the messages consumed is exactly the number the
       // maximum allowed because of eventual consistency in SQS
       expect(messagesConsumed).toBeGreaterThan(0);
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply example/spec/parallel/startSF/startSfSpec.js]: So the amount of messages it would consume (up to the message limit passed to the Lambda) is unpredictable here as well. Before, I did have this written as:

`expect(messagesConsumed).toBeLessThanOrEqual(queueMaxExecutions)`

However, there are some cases where the Lambda consumer starting executions runs longer than the lifespan of these executions (5 seconds), so the semaphore is decremented for executions that finish and the consumer is able to start more executions. This happens pretty infrequently, but it does happen. 

Given these kind of possible timing inconsistencies, I decided that the only important time to enforce the maximum execution limit is respected is when we count the number of running executions. What do you think?

File: example/spec/parallel/startSF/startSfSpec.js (hunk 4 above)
Comment (ifestus) [reply example/spec/parallel/startSF/startSfSpec.js]: :doh: good deal.

File: example/spec/parallel/startSF/startSfSpec.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ifestus) [reply example/spec/parallel/startSF/startSfSpec.js]: Enforcing that the number of running executions respects an execution limit sounds like the right path to take.
Taking a small step back, what is the intent of the `expect(messagesConsumed).toBeGreaterThan(0)` line?

[the remark this answers is not in the document]
Comment (markdboyd) [reply example/spec/parallel/startSF/startSfSpec.js]: Yeah, it's clearly not testing very much other than that at least **some amount** messages were consumed by the lambda from the queue. But given the inability to predict how many or guarantee that the number will be less than the executions max, the test can't really be much more precise

[the remark this answers is not in the document]
Comment (ifestus) [reply example/spec/parallel/startSF/startSfSpec.js]: Is there any value in changing that `0` to a higher value? Correct me if I'm wrong, but we should be able to say with certainty that _at least_ the maximum executions number of messages should be read off of the queue and started, right?

[the remark this answers is not in the document]
Comment (markdboyd) [reply example/spec/parallel/startSF/startSfSpec.js]: Unfortunately, we cannot say that with certainty, because SQS is eventually consistent and reading from it right after you've put messages in the queue may not give you that exact number of messages. This is also why the test above only checks that the number of messages consumed is greater than 0

[the remark this answers is not in the document]
Comment (ifestus) [reply example/spec/parallel/startSF/startSfSpec.js]: Sounds good to me. Thanks for walking through that with me. The final suggestion I have, then, would be to change the string of this `it` block from `consumes the right amount of message` to something like `consumes some number of messages`. This isn't a blocking change, but it makes the test more readable to me.
